### ISSUE TITLE
Improve release flow for libs

### ISSUE THREAD

**Issue by hjalmers** [sebgroup/green#264]
We currently use https://github.com/ng-easy/platform/tree/main/libs/builders/src/semantic-release to run semantic release with nx. There are som other options too but none seem to provide an easy solution out of the box. One hurdle to overcome involves determining which commits belong to which project. Even though affected files belong to a certain project it seems like semantic release currently looks at all commits when determining next version and when generating the change logs. Maybe enforcing scopes eg. fix(angular): foo or feat(chlorophyll): bar would be a good starting point and then write a custom commit analyser and change log parser...

**Comment by hjalmers** [comment]
I've been doing some more digging and thinking.

Goal and taget setup

### Release of lib and changelog generation
- **Determined by commit message type:** We should enforce a commit message format e.g. angular conventional commits (which we're currently using) and determine kind of release based on commit type i.e (fix, feat etc.).
- **Determined by touched files:** Which lib to release should be determined by looking at touched files. Commits touching files in source folder of `lib/chlorophyll` e.g. `fix(foo): bar lorem ipsum` should create a new patch release for chlorophyll and include the commit message in the changelog for that particular lib.
- **Multiple libs in one commit:** Ideally a commit should just affect one lib at a time, changes to multiples libs (e.g. react and extract) should preferably be split into two separate commits. However if the commit affects multiple libs at the same time and the change can be described and convey meaning to multiple libs using a single commit message it should be okay to just do one single commit instead of two (or more). The key rule here is that the commit will create a release for each lib touched by the commit and the same (single) commit message will be added to each changelog (one per lib).
- **Individual versioning:** Since our monorepo and our libs define some packages which have no relationship or correlation, it makes sense to release them individually, i.e we probably don't want to bump green-angular just because react releases a new version or the other way around as it would result in a lot of "unnecessary" releases and noice.  
-  **Managing local dependencies:**  If one package has a dependency on another, eg. green-angular depends on chlorophyll then a release of chlorophyll should trigger a patch release of green-angular were the dependency to chlorophyll as been bumped.

Currently semantic release has no official support for monorepos but there are some third-party options and plugins which try, to varying degree and success, to add support for some of the things listed above. [This article](https://dev.to/antongolub/the-chronicles-of-semantic-release-and-monorepos-5cfc) describes part of the history and some of the workarounds and options available. We currently use another one which seems to have a few issues or perhaps we're using it or have configured it correctly. I suspect that our commit history and the way we manage pull request are affecting the results. We could also have a go a [this one](https://github.com/qiwi/multi-semantic-release) but I still think we need to enforce and restrict how we manage pull requests. Squashing multiple commits from a branch will result in a single commit i.e one message for all files affect by the PR. I assume we don't want to squash commits unless they actually belong together and should generate a single release for affect files with a single commit message.

Another option to using semantic-release and automating releases depending on commit messages could be using something like [auto](https://intuit.github.io/auto/) or writing something custom.

**Comment by mario-subo** [comment]
Thanks for the explanation and all the references @hjalmers . I vote for `auto` as it seems to be the simplest and least hacky solution.

**Comment by hjalmers** [comment]
I've implemented this now: https://github.com/jscutlery/semver it's one of the more popular "plugins" for nx to handle versioning of libs. It's not perfect and it doesn't come with all the bells and whistles that you get when using something relying on https://semantic-release.gitbook.io/semantic-release/, however - it's fast and easy enough to setup and configure.

I've written a custom executor to bump dependencies to other green libs (note that the order in which they are versioned is important as they depend on each other).

So far I've just set it up so that we can manually trigger a prerelease of all libs, and since they're all published at the same time now using beta as pre-release tag, they'll use the same version number and patches, features etc. will result in the same version increment e.g 1.0.0-beta.6 👉  1.0.0-beta.7.  Once we're happy to go live with v1.0.0 we should either create a new release flow or alter it so that version and release is determined by commit type and if a lib contains changes.

I've also disabled the ability squash pull requests as this prevent us from tying a specific commit to a library if the commit contains changes to multiple libs. Eg.

- fix(charts): foo bar (touched files in chart lib)
- feat(charts): add ability to foo (touched files in chart lib)
- fix(card): use correct margin (touched files in chlorophyll lib)

If squashed, when analysed we'd just see:

- fix(charts): foo bar (touched files in chart and chlorophyll lib) (or whatever commit message we use for the squash commit)

Resulting in a single change log post which probably is not what we want in this case.
